src/chainstate/stacks/index/storage.rs

```diff
@@ -321,8 +632,221 @@ impl<T: MarfTrieId> TrieRAM<T> {
         Ok(())
     }
 
+    /// Calculate the MARF root hash from a trie root hash.
+    /// This hashes the trie root hash with a geometric series of prior trie hashes.
+    fn calculate_marf_root_hash(
+        &mut self,
+        storage: &mut TrieStorageTransaction<T>,
+        root_hash: &TrieHash,
+    ) -> TrieHash {
+        let (cur_block_hash, cur_block_id) = storage.get_cur_block_and_id();
+
+        storage.data.set_block(self.block_header.clone(), None);
+
+        let marf_root_hash = Trie::get_trie_root_hash(storage, root_hash)
+            .expect("FATAL: unable to calculate MARF root hash from moved TrieRAM");
+
+        test_debug!("cur_block_hash = {}, cur_block_id = {:?}, self.block_header = {}, have last extended? {}, root_hash: {}, trie_root_hash = {}", &cur_block_hash, &cur_block_id, &self.block_header, storage.data.uncommitted_writes.is_some(), root_hash, &marf_root_hash);
+
+        storage.data.set_block(cur_block_hash, cur_block_id);
+
+        marf_root_hash
+    }
+
+    /// Calculate and store the MARF root hash, as well as any necessary intermediate nodes.  Do
+    /// this only for deferred hashing mode.
+    fn inner_seal_marf(
+        &mut self,
+        storage_tx: &mut TrieStorageTransaction<T>,
+    ) -> Result<TrieHash, Error> {
+        // find trie root hash
+        debug!("Calculate trie root hash");
+        let root_trie_hash = self.calculate_node_hashes(storage_tx, 0)?;
+
+        // find marf root hash -- the hash of the trie root node hash, and the hashes of the
+        // geometric series of ancestor tries.  Because the trie is already in the process of
+        // being flushed, we have to temporarily reinstate its data into `storage_tx` so we can
+        // use it to walk down the various MARF paths needed to query ancestor tries.
```

Comment:
I'm not sure I follow why the trie already being flushed makes "reinstatement" necessary for finding the marf root hash. What breaks in the normal MARF walk if you tried to `calculate_marf_root_hash` on `self` as it is?

Reply:
The call to `inner_flush` has this:

```rust
        if let Some((bhh, trie_ram)) = self.data.uncommitted_writes.take() {
            trace!("Buffering block flush started.");
            let mut buffer = Cursor::new(Vec::new());
            trie_ram.dump(self, &mut buffer, &bhh)?;
```

Because `self.data.uncommitted_writes` will be `None` at the time of the call to `trie_ram.dump()`, it will be impossible to carry out a MARF walk.  The uncommitted TrieRAM contains the root node from which the walks will begin.